cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -660,6 +663,100 @@ struct CastFunctor<T, DictionaryType,
   }
 };
 
+// ----------------------------------------------------------------------
+// String to Number
+
+template <typename O>
+struct CastFunctor<O, StringType,
+                   typename std::enable_if<std::is_base_of<Number, O>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using out_type = typename O::c_type;
+    StringArray input_array(input.Copy());
+
+    if (input_array.null_count() > 0) {
+      std::stringstream ss;
+      ss << "Failed to cast NA into " << output->type->ToString();
+      ctx->SetStatus(Status(StatusCode::SerializationError, ss.str()));
+      return;
+    }
+
+    auto out_data = GetMutableValues<out_type>(output, 1);
+
+    std::function<out_type(const std::string&)> cast_func;
+    if (output->type->id() == Type::INT8 || output->type->id() == Type::UINT8) {
+      cast_func = [](const std::string& s) {
+        return boost::numeric_cast<out_type>(boost::lexical_cast<int>(s));
+      };
+    } else {
+      cast_func = [](const std::string& s) { return boost::lexical_cast<out_type>(s); };
```

Comment:
I think C++11 Lambdas actually incur more overhead than an inlined function. We should instead introduce an auxiliary numeric cast functor that does this switch at compile-time (resulting in an inlined function in the inner loop for all possible types) rather than runtime